src/VGrid.cu

```diff
@@ -0,0 +1,554 @@
+		   int                         nbatch,
+		   int                         nchan,
+		   const int* __restrict__     x,
+		   const int* __restrict__     y,
+		   const int* __restrict__     z,
+		   const OutType* __restrict__ kernels,
+		   const InType* __restrict__  d_in,
+		   OutType*                    d_out) {
+        
+	/// Thread and Block Indexes and Dimensions
+
+    int bid_x = blockIdx.x, bid_y = blockIdx.y, bid_z = blockIdx.z ;
+    int blk_x = blockDim.x, blk_y = blockDim.y;
+    int grid_y = gridDim.y, grid_z = gridDim.z ;
+    int illum_x = threadIdx.x, tid_y = threadIdx.y ;
+
+    int vi_s = (bid_y+bid_x*grid_y)*grid_z*blk_y*npol ;
+    int grid_s = (bid_y+bid_x*grid_y)*npol*gridsize*gridsize;
+
+    /// Dynamic allocation of shared memory
+
+    extern __shared__ float shared[];
+    
+    float* xdata = shared;
+    float* ydata = xdata + blk_y * npol;
+    
+    /// Store X- and Y- Locations of Antennas within a thread-block in shared memory
+
+    #pragma unroll
+    for(int kk=0;kk<npol;kk++)
+    {
+          xdata[tid_y*npol + kk] = x[vi_s + npol *(bid_z*blk_y+tid_y)+kk];	
+	  ydata[tid_y*npol + kk] = y[vi_s + npol *(bid_z*blk_y+tid_y)+kk];
+    }  
+    __syncthreads();
+
+       int myU = illum_x% maxsupport; int myV = illum_x / maxsupport;
+       float grid_point_u = myU; float grid_point_v = myV;
+       OutType sum = OutType(0.0, 0.0);
+  
+       #pragma unroll
+       for(int pol = 0; pol < npol;pol++) 
+       {
+                 
+	    float xl = xdata[tid_y*npol+pol]; 
+	    float yl = ydata[tid_y*npol+pol];
+            // Determine convolution point. This is basically just an
+            // optimised way to calculate.
+            float myConvU = 0.0; 
+	    float myConvV = 0.0;
+            if( maxsupport > 1 ) 
+	    {
+                myConvU = (int(xl) - myU) % maxsupport; myConvV = (int(yl) - myV) % maxsupport;    
+                if (myConvU < 0) myConvU += maxsupport; if (myConvV < 0) myConvV += maxsupport;
+            } 
+            // Determine grid point. Because of the above we know here that
+           float myGridU = xl + myConvU; float myGridV = yl + myConvV;
+           
+       	   // If grid-point changes for a given illumination pattern 
+
+          if (!(myGridU == grid_point_u && myGridV == grid_point_v)) 
+	  //if(!((myGridU<grid_point_u+0.5)&&(myGridU>=grid_point_u-0.5)&&(myGridV<grid_point_v+0.5)&&(myGridU>=grid_point_u-0.5)))
+	   { // Atomic add to grid is now removed from this kernel after careful examining and consideration for race-conditions
+               if( grid_point_u >= 0 && grid_point_u < gridsize && \
+                    grid_point_v >= 0 && grid_point_v < gridsize ) 
+	        {
+                       d_out[grid_s + pol*gridsize*gridsize + gridsize*int(grid_point_v) + int(grid_point_u)].x+= sum.x;
+                       d_out[grid_s + pol*gridsize*gridsize + gridsize*int(grid_point_v) + int(grid_point_u)].y+= sum.y;   
```

Comment:
I (still) don't understand this step. As far as I can tell, at this point in the code, `sum` is always `(0, 0)`.
It's initialized as `(0, 0)` at the start of each `kk` iteration, and we shouldn't be entering this `if` conditional from one polarization to the next.
What is this supposed to be doing?